array/builder.go

```diff
@@ -38,54 +32,64 @@ func (b *StringBuilder) Retain() {
 }
 func (b *StringBuilder) Release() {
 	if atomic.AddInt64(&b.refCount, -1) == 0 {
-		if b.buffer != nil {
-			b.buffer.Release()
+		if b.indices != nil {
+			b.indices.Release()
+			b.indices = nil
 		}
-		if b.builder != nil {
-			b.builder.Release()
+		if b.values != nil {
+			b.values.Release()
+			b.values = nil
 		}
 	}
 }
 func (b *StringBuilder) Len() int {
-	if b.builder != nil {
-		return b.builder.Len()
+	if b.hydratedArray {
+		return b.values.Len()
 	}
-	return b.len
+	return b.indices.Len()
 }
 func (b *StringBuilder) Cap() int {
-	if b.builder != nil {
-		return b.builder.Cap()
-	}
-	if b.cap > b.len {
-		return b.cap
+	if b.hydratedArray {
+		return b.values.Cap()
 	}
-	return b.len
+	return b.indices.Cap()
 }
 func (b *StringBuilder) NullN() int {
-	if b.builder != nil {
-		return b.builder.NullN()
+	if b.hydratedArray {
+		return b.values.NullN()
 	}
-	return 0
+	return b.indices.NullN()
 }
 
 func (b *StringBuilder) AppendBytes(buf []byte) {
-	if b.builder != nil {
-		b.builder.Append(buf)
-		return
-	}
-	if b.len == 0 {
-		b.buffer = memory.NewResizableBuffer(b.mem)
-		b.buffer.Resize(len(buf))
-		copy(b.buffer.Bytes(), buf)
-		b.len = 1
-		return
-	}
-	if bytes.Equal(b.buffer.Bytes(), buf) {
-		b.len++
-		return
+	if !b.hydratedArray {
+		if b.values.Len() == 0 {
+			b.values.Append(buf)
+			b.indices.Append(0)
+			return
+		}
+		if bytes.Equal(buf, b.values.Value(0)) {
+			b.indices.Append(0)
+			return
+		}
+		// Need to add a new value to the values array, that means we
+		// need to hydrate it.
+		b.hydratedArray = true
+		indices := b.indices.NewInt32Array()
+		values := b.values.NewBinaryArray()
+		b.values.Reserve(indices.Len())
+		b.values.ReserveData(values.ValueLen(0) * indices.Len())
+		for i := 0; i < indices.Len(); i++ {
+			if indices.IsNull(i) {
+				b.values.AppendNull()
+			} else {
+				b.values.Append(values.Value(int(indices.Value(i))))
+			}
+		}
+		values.Release()
+		indices.Release()
```

Comment:
Feels like I'm back in my C++ days 😄

Reply:
That is because the go API is modelled off of the C++ one.